plugin/ui/completion.py

```diff
@@ -274,3 +295,94 @@ def show(self) -> None:
     @staticmethod
     def hide(view: sublime.View) -> None:
         mdpopups.hide_popup(view)
+
+
+class _PhantomCompletion(_BaseCompletion):
+    COPILOT_PHANTOM_COMPLETION = "copilot_phantom_completion"
+    PHANTOM_TEMPLATE = """
+    <body id="copilot-completion">
+        <style>
+            body {{
+                color: #808080;
+            }}
+
+            .copilot-completion-line {{
+                margin-top: {line_padding_top}px;
+                margin-bottom: {line_padding_bottom}px;
+            }}
+
+            .copilot-completion-line.first {{
+                margin-top: 0;
+            }}
+        </style>
+        {body}
+    </body>
+    """
+    PHANTOM_LINE_TEMPLATE = '<div class="copilot-completion-line {class_name}">{content}</div>'
+
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+
+        self._phantom_set = self._get_phantom_set(view, True)
+
+    @classmethod
+    def _get_phantom_set(cls, view: sublime.View, init: bool = False) -> Optional[sublime.PhantomSet]:
+        view_id = view.id()
+        phantom_set = _phantom_sets_per_view.get(view_id, None)
+
+        if init:
+            phantom_set = sublime.PhantomSet(view, cls.COPILOT_PHANTOM_COMPLETION)
+            _phantom_sets_per_view[view_id] = phantom_set
+
+        return phantom_set
+
+    def normalize_phantom_line(self, line: str) -> str:
+        # return re.sub(r"( {2,})", lambda m: "&nbsp;" * len(m.group(0)), html.escape(line))
+        return (
+            html.escape(line).replace(" ", "&nbsp;").replace("\t", "&nbsp;" * self.view.settings().get("tab_size", 4))
+        )
+
+    def _build_phantom(
+        self, lines: Union[str, List[str]], begin: int, end: Optional[int] = None, inline: bool = True
+    ) -> sublime.Phantom:
+        body = (
+            "".join(
+                self.PHANTOM_LINE_TEMPLATE.format(
+                    class_name=("first" if index == 0 else ""), content=self.normalize_phantom_line(line)
+                )
+                for index, line in enumerate(lines)
+            )
+            if isinstance(lines, list)
+            else self.normalize_phantom_line(lines)
+        )
+
+        return sublime.Phantom(
+            sublime.Region(begin, begin if end is None else end),
+            self.PHANTOM_TEMPLATE.format(
+                body=body,
+                line_padding_top=int(self.view.settings().get("line_padding_top", 0)) * 2,  # TODO: play with this more
+                line_padding_bottom=int(self.view.settings().get("line_padding_bottom", 0)) * 2,
+            ),
+            sublime.LAYOUT_INLINE if inline else sublime.LAYOUT_BLOCK,
+        )
+
+    def show(self) -> None:
+        completion = self.completion_manager.current_completion
+        assert completion
+
+        head, *tail = completion["displayText"].split("\n")
```

Comment:
`str.partition` is perfect for this use. It should be faster and more memory efficient (not important here though 😄 )

Reply:
Unfortunately `str.partition` won't work, because I need tot split text into an array, so I can render each line separately

Reply:
> Unfortunately `str.partition` won't work here, because I need to split the text into a list, so I can render each line separately

So you handle the second line in a different way with the third line? I don't expect that. I think only the first and the rest are different.

Reply:
Basically I need a list of lines. 
The first line is rendered as an inline phantom and all other lines are rendered as a block phantom but each line needs to be wrapped in `<div class="copilot-completion-line>`, that's why I need the `split`

Reply:
![image](https://user-images.githubusercontent.com/6594915/182427284-535c56c8-06b6-4866-8820-47db782b63c5.png)

From my experiment, text in a phantom also auto follows `line_padding_top` and `line_padding_bottom`. There is no need for CSS.

Reply:
It follows only in the inline phantom(first line), and I don't add padding there.  
But, unfortunately, it doesn't follow in the block phantom that is displayed underneath

Reply:
> But, unfortunately, it doesn't follow in the block phantom that is displayed underneath

ah, I see that now...